feat(mvux): Added Advanced page 🚧 (#1398)
unoplatform/uno.extensions: doc/Overview/Mvux/Advanced/Pagination.md

When the user requests a new page, the current page index state is updated, thereby updating the dependent collection List-Feed.

Using the example started in [incremental loading above](#incremental-loading), we'll add another method to the service, which will disclose to the View how many items are there in total. Getting a count of items is more efficient than enumerating all entries. This is necessary to identify the total number of pages we have.

### Service

Here's the methods added to the service (*PeopleService.cs*):

```csharp
public IFeed<uint> PageCount =>
    Feed.Async(async (ct) => await PeopleService.GetPageCount(DefaultPageSize, ct));
```

The signature of `GetPageCount` should also be added to the `IPeopleService` interface:

```csharp
ValueTask<uint> GetPageCount(uint pageSize, CancellationToken ct);
```

### Model

Let's expand the Model with the following:

```csharp
public partial record PeopleModel(IPeopleService PeopleService)
{
    public IListFeed<Person> PeopleAuto ...// will not be used in this example

>     public IFeed<uint> PageCount =>

Reviewer: Please rework this example to remove the GetPageCount on the service. The total number of items/pages should come as part of a call to fetch each page of data

Author: We have 3 examples in the pagination project. The middle example implements manual pagination. The page count is required in this example.

---
uid: Overview.Mvux.Advanced.Pagination
---

# Pagination

There are several ways to paginate data.

> [!NOTE]  
> The source-code for the sample app demonstrated in this section can be found [here](https://github.com/unoplatform/Uno.Samples/tree/master/UI/MvuxHowTos/PaginationPeopleApp).

## Incremental loading

The easiest and most straight-forward is use the built-in incremental loading functionality that some controls (e.g. `ListView`, `GridView`) offer via the [`ISupportIncrementalLoading`](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.data.isupportincrementalloading) interface the paginated Feed implements.

For the Pagination example we'll also use the *PeopleApp* example used above in [Selection](xref:Overview.Mvux.Advanced.Selection), but you can find the full code [here](https://github.com/unoplatform/Uno.Samples/tree/master/UI/MvuxHowTos/PaginationPeopleApp).

### Service

Let's change the `GetPeopleAsync` method in the *PeopleService.cs* file to support pagination:

```csharp
namespace PaginationPeopleApp;

public partial record Person(string FirstName, string LastName);

public interface IPeopleService
{

>     ValueTask<IImmutableList<Person>> GetPeopleAsync(uint pageSize, uint pageIndex, CancellationToken ct);

Reviewer: Do these need to be uint? If not, please just use int as this is what most devs will default to

Author: Added comment explaining why `uint` is used.

    ValueTask<uint> GetPageCount(uint pageSize, CancellationToken ct);
}

public class PeopleService : IPeopleService
{
    public async ValueTask<IImmutableList<Person>> GetPeopleAsync(uint pageSize, uint pageIndex, CancellationToken ct)
    {
        // convert to int for use with LINQ

>         var (size, index) = ((int)pageSize, (int)pageIndex);

Author: They will have to be converted either here or in the model, as uint is the type originated from the pagination source.

>         var (size, index) = ((int)pageSize, (int)pageIndex);

Author: @dr1rrb please see this: https://github.com/unoplatform/uno.extensions/issues/1437.

>         var (size, index) = ((int)pageSize, (int)pageIndex);

Author: @nickrandolph I think for the moment let's just leave it as is. If it'll be changed to ints, I'll obviously be happy to cut all the casting clutter off.

>         var (size, index) = ((int)pageSize, (int)pageIndex);

Author: @dr1rrb What do you mean, is LINQ Skip and Take fake APIs?

> [!NOTE]  
> The source-code for the sample app demonstrated in this section can be found [here](https://github.com/unoplatform/Uno.Samples/tree/master/UI/MvuxHowTos/PaginationPeopleApp).

## Offset pagination

Offset pagination is controlled via a State that stores the current page index in the Model, and the List-Feed depending on it, using [the `Select` operator](xref:Overview.Mvux.Feeds#select-or-selectasync).  
When the user requests a new page, the current page index state is updated, thereby updating the dependent collection List-Feed.

Using the example started in [incremental loading above](#incremental-loading), we'll add another method to the service, which will disclose to the View how many items are there in total. Getting a count of items is more efficient than enumerating all entries. This is necessary to identify the total number of pages we have.

### Service

Here's the method added to the service (*PeopleService.cs*):

```csharp

> public async ValueTask<int> GetPageCount(uint pageSize, CancellationToken ct) =>

Reviewer: This isn't a great example. Most services that serve paged content will return total as part of the paged result, so you don't need to make a separate call to get the page/item count

Author: 1. We want to know the total number of pages before we start to determine if the "next" button should be available at all
2. We're not adding any new data in our example, so the count is static throughout the app, so why retrieve the count on each pagination request?

Reviewer: I'm just saying that returning total number of items in the first request is more common than having a service call that returns the number of pages/items, so I'd prefer to see that as the example
In fact, I've never seen a case where I've had to request the number of items/pages and then make a request for the items

Author: how would i know if there is a another page and set the "next" button enabled state accordingly?
we need to have a feed that keeps this info

Author: I've redone the pagination page. This method is now only used in "manual pagination" section.

        await CurrentPage.Set((uint)desiredPage, ct);
    }
}
```

In the code above, whenever `CurrentPage` changes, `PeopleManual` will update its data accordingly.  
MVUX's code generator generates an async command that is bound to from the View, when the user clicks the designated button to move to either previous or next page, the `Move` method will be invoked with the `direction` argument set to either 1 or -1 added to the current page index.

You may be wondering where the `currentPage` and `pageCount` arguments are populated from, this is another MVUX magic trick where when the parameters type and name matches a Feed or a State in the current Model, their latest value are provided as an argument to the executed Command.  
Read onward on [Creating commands](xref:Overview.Mvux.Advanced.Commands#creating-commands) to learn more.

### View

Replace the `ListView` from the previous example with this one:

```xml
<ListView Grid.Column="2" x:Name="manual" ItemsSource="{Binding PeopleManual}"
          ItemTemplate="{StaticResource PersonDataTemplate}" Header="Load single page on demand">
    
    <ListView.Footer>
        <StackPanel Orientation="Horizontal" Spacing="3">

>             <Button Content="Previous" Command="{Binding Move}" VerticalAlignment="Center">

Reviewer: ```suggestion
            <Button Content="Previous" Command="{Binding Move}" VerticalAlignment="Center" CommandParameter="1">
```
This would be much nicer!

Author: You bet I've tried that before writing my clunky explicit int declaration 🤢

                </Button.CommandParameter>
            </Button>

            <TextBlock Text="{Binding CurrentPage}" VerticalAlignment="Center" />

            <Button Content="Next" Command="{Binding Move}" VerticalAlignment="Center">
                <Button.CommandParameter>
                    <x:Int32>1</x:Int32>
                </Button.CommandParameter>
            </Button>

        </StackPanel>
    </ListView.Footer>
</ListView>
```

Two buttons were added to the `ListView`'s footer which are bound to the `Move` command, and the `TextBlock` in-between them will display the current page via the `CurrentPage` State it's bound to.  
A `CommandParameter` of either *1* or *-1* has been set to Next and Previous buttons respectively.

![](../Assets/PaginationManual-1.jpg)

> [!TIP]  
> The generated `Move` command can be seen in the generated `BindablePeopleModel` as well, follow the instructions [here](xref:Overview.Mvux.Advanced.InspectGeneratedCode).

When running the app, the first page will be loaded and await the users input to navigate to other pages which will be loaded on-demand:

![](../Assets/PaginationManual-2.gif)

### The `PageRequest` type

> 

Author: It's explaining the `PageRequest` type, which is an integral part of MVUX Pagination.

> 

Author: Generally subscribing to the selection using the ForEach operator.
@dr1rrb I've sent you a comment I need more advice on this.

## Keyset pagination with a cursor

There are several caveats in using `Skip` and `Take` (Offset pagination) with an arbitrary page size multiplied by the page number.

- When we skip data records, the database might still have to process some the skipped records on its way to the desired ones.
- If any updates have been applied to the records preceding the currently displayed page, and then the user moves to the next or previous page, there might be inconsistencies in showing the subsequent data, some of the entries might be skipped or shown twice.

An alternative way to paginate data is by using a cursor that points to a specific record and take then number of desired records in a page onwards.  
This is referred to as 'keyset pagination' or 'seek-based pagination'.

> To utilize this pagination style, MVUX provides another `ListFeed` factory overload, the `AsyncPaginatedByCursor`.

Author: I added "jump to example" link.

### Model

```csharp
using Uno.Extensions.Reactive;

namespace PaginationPeopleApp;

public partial record PeopleModel(IPeopleService PeopleService)
{
    const uint PageSize = 20;

    public IListFeed<Person> PeopleAuto =>
        ListFeed.AsyncPaginated(async (PageRequest pageRequest, CancellationToken ct) =>
            await PeopleService.GetPeopleAsync(pageSize: PageSize, pageIndex: pageRequest.Index, ct));
}
```

The `AsyncPaginated` method generates a `ListFeed` that supports pagination.

Whenever the user scrolls down to see additional data and is hitting the end of the collection displayed in a `ListView`, the pagination List-Feed is automatically triggered with a page request.  
The parameter of `AsyncPaginated`, is a delegate taking in a [`PageRequest`](#the-pagerequest-type) value and a `CancellationToken` and returning an `IListFeed<T>` where `T` is `Person` in our case. This delegate is invoked when a page request comes in.

As you can see, we are sending the `Index` property of the incoming `pageRequest` argument to determine what page we're positioned.  

> `PageSize` refers to a constant value of `20`, but the `PageRequest` also has a `DesiredSize` property which the `ListView` can set according to its capacity. So essentially you can substitute `pageSize: PageSize` with `pageSize: pageRequest.DesiredSize` which is a nullable `uint`, and is populated by the `ListView`. We used a constant `PageSize` for the sake of this demonstration.

Author: Why is it null? How can this number be determined in the first call?

### View

```xml
<Page 
    x:Class="PaginationPeopleApp.MainPage"
	xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
	xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:PaginationPeopleApp"
    xmlns:mvux="using:Uno.Extensions.Reactive.UI">

    <Page.Resources>
        <DataTemplate x:Key="PersonDataTemplate" x:DataType="local:Person">
            <StackPanel Orientation="Horizontal" Spacing="2">
                <TextBlock Text="{x:Bind FirstName}"/>
                <TextBlock Text="{x:Bind LastName}"/>
            </StackPanel>
        </DataTemplate>
    </Page.Resources>

    <ListView ItemsSource="{Binding PeopleAuto}" ItemTemplate="{StaticResource PersonDataTemplate}"

>         DataFetchSize="20" Header="Next page auto-loading when on available capacity or scrolled"/>

Reviewer 2: We should not set the `DataFetchSize`, I'm not even sure that is supported by Uno.

</Page>
```

In addition, make sure the `DataContext` of the `Page` is set to the generated bindable model called `BindablePeopleModel` (*MainPage.xaml.cs*):

```csharp
public MainPage()
{
    this.InitializeComponent();

    this.DataContext = new BindablePeopleModel(new PeopleService());
}
```

> [!TIP]  
> You can inspect the generated code by either placing the cursor on the word `BindablePeopleModel` and hitting <kbd>F12</kbd>, see other ways to inspect the generated code [here](xref:Overview.Mvux.Advanced.InspectGeneratedCode).

As you can see, there's nothing special in the XAML code as MVUX is taking advantage of the tools already implemented with the `ListView`.

- When the page load, the first 20 items are loaded (after a delay of one second - as simulated in the service).

> - If there's available space in the `ListView`, the next batch of 20 items will be requested from the Feed and loaded from the service thereafter.

Reviewer 2: I would encourage the use of the `DesiredSize` instead of relying on a hard coded page size of 20.

Author: I chose to write 20 because the default value is 76. Like 4 pages ahead. So the demonstration is a bit lame as it's not evident data is loading.

Reviewer 2: The 76 depends of the size of the `ListView` and the size of the items. For you sample you can just increase the height of your items, so the `ListView` will requests smaller pages.

- When the user scroll down and hits the bottom of the `ListView`, the next 20 items will be requested.
- This behavior will follow as the user scrolls down and chases the service for more data until all items have been loaded.

Here's what the app renders like:

![](../Assets/PaginationIncrementalLoading.gif)

> [!TIP]  

> > You can use these members of the `ListView` to control data loading: [`DataFetchSize`](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.listviewbase.datafetchsize), [`IncrementalLoadingThreshold`](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.listviewbase.incrementalloadingthreshold), and [`IncrementalLoadingTrigger`](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.listviewbase.incrementalloadingthreshold).

Reviewer 2: Not sure all of this is supported by Uno, did you validated? I think that we should not promote them if they are not.

Using the example started in [incremental loading above](#incremental-loading), we'll add another method to the service, which will disclose to the View how many items are there in total. Getting a count of items is more efficient than enumerating all entries. This is necessary to identify the total number of pages we have.

### Service

Here's the method added to the service (*PeopleService.cs*):

```csharp
public async ValueTask<int> GetPageCount(uint pageSize, CancellationToken ct) =>
    (uint)Math.Ceiling(GetPeople().Length / (double)pageSize);
```

The signature of `GetPageCount` should also be added to the `IPeopleService` interface:

```csharp
ValueTask<int> GetPageCount(uint pageSize, CancellationToken ct);
```

### Model

Let's expand the Model with the following:

```csharp
public partial record PeopleModel(IPeopleService PeopleService)
{
    public IListFeed<Person> PeopleAuto ...// will not be used in this example

    const uint PageSize = 20;
        
    public IFeed<int> PageCount =>
        Feed.Async(async (ct) => await PeopleService.GetPageCount(PageSize, ct));

    public IState<uint> CurrentPage => State.Value(this, () => 1u);

>     public IListFeed<Person> PeopleManual =>
>         CurrentPage.SelectAsync(async (currentPage, ct) =>
>             // currentPage argument as index based
>             await PeopleService.GetPeopleAsync(pageSize: PageSize, pageIndex: currentPage - 1, ct))
>         .AsListFeed();

Reviewer 2: ```suggestion
    public IListFeed<Person> PeopleManual => CurrentPage
        // currentPage argument as index based
        .SelectAsync(async (currentPage, ct) => await PeopleService.GetPeopleAsync(pageSize: PageSize, pageIndex: currentPage - 1, ct))
        .AsListFeed();
```

Much more readable IMO. My rule is either no wrapping (if line is short), either add a new line for each operator.

Author: @dr1rrb

> Much more readable IMO. My rule is either no wrapping (if line is short), either add a new line for each operator.

Actually my rule when using functional / fluent **every** operator in new line regardless of length of line, and the class itself as well.

How about:

```suggestion
    public IListFeed<Person> PeopleManual => 
        CurrentPage
            // currentPage argument as index based
            .SelectAsync(async (currentPage, ct) => await PeopleService.GetPeopleAsync(pageSize: PageSize, pageIndex: currentPage - 1, ct))
            .AsListFeed();
```

BTW, In my personal code I'd do:

```c#
    public IListFeed<Person> PeopleManual => 
        CurrentPage
        // currentPage argument as index based
        .SelectAsync(async (currentPage, ct) =>
            await PeopleService
                  .GetPeopleAsync(
                      pageSize: PageSize, 
                      pageIndex: currentPage - 1, 
                      ct))
        .AsListFeed();
```
I know it's a bit extreme.

Author: > Much more readable IMO. 

I disagree. `CurrentPage` appearing on the right side of the lambda operator is pretty much invisible.

>  My rule is either no wrapping (if line is short), either add a new line for each operator.

It's also my rule, but I'm also placing the container in a new line. I'll update code accordingly.

Reviewer 2: The main point here is to wrap before the `.SelectAsync` and keep the inner delegate on a single line as it's a method body. I would wrap a delegate only if multi-line, i.e. with curly brackets.

For the `CurrentPage` not being visible, I don't really agree, for me it makes clearer that the `PeopleManual` is derived from the `CurrentPage` (`PeopleManual => CurrentPage` being on a single line) on which we apply some rules (that are on lines below).

My rule for expression-bodied members is to keep them on a single line for property:
```csharp
public string A => _a.ToString();`
```
and wrap before the arrow for methods:
```csharp
public string GetA()
    => _a.ToString();
```

But all of those is only personal opinion (and I against strong rules for those!) Every dev as it's own way to think / work and as long as the alignment makes sense I'm fine with it :)

|Property|Description|
|---|---|
|Index|The index of the page to be loaded.|
|CurrentCount|This is the total number of items currently in the list.|
|DesiredSize|The desired number of items for the current page, if any.<br/><br/>This is the desired number of items that the view requested to load.<br/>It's expected to be null only for the first page.<br/>Be aware that this might change between pages (especially is user resize the window), DO NOT use in `source.Skip(page.Index * page.DesiredSize).Take(page.DesiredSize)`.<br/>Prefer to use the `CurrentCount` property, e.g. `source.Skip(page.CurrentCount).Take(page.DesiredSize)`.|

## Keyset pagination with a cursor

There are several caveats in using `Skip` and `Take` (Offset pagination) with an arbitrary page size multiplied by the page number.

- When we skip data records, the database might still have to process some the skipped records on its way to the desired ones.
- If any updates have been applied to the records preceding the currently displayed page, and then the user moves to the next or previous page, there might be inconsistencies in showing the subsequent data, some of the entries might be skipped or shown twice.

> An alternative way to paginate data is by using a cursor that points to a specific record and take then number of desired records in a page onwards.  

Reviewer 3: ```suggestion
An alternative way to paginate data is by using a cursor that points to a specific record and takes the number of desired records in a page onwards.  
```

This is referred to as 'keyset pagination' or 'seek-based pagination'.

To utilize this pagination style, MVUX provides another `ListFeed` factory overload, the `AsyncPaginatedByCursor`.

The signature of this method is:

```csharp
public static IListFeed<T> AsyncPaginatedByCursor<TCursor>(TCursor firstPage, GetPage<TCursor, T> getPage)
```

where the `GetPage` type is a delegate with the following pattern:

```
public delegate ValueTask<PageResult<TCursor, TItem>> GetPage<TCursor, TItem>(TCursor cursor, uint? desiredPageSize, CancellationToken ct);
```

The callback delegate takes two parameters and returns two.

The two input parameters are:

> - A cursor to the first page, in other words the item this page starts with. The service then starts by ordering the data and searching for the first item, then uses the second parameter:

Reviewer 3: ```suggestion
- A cursor to the first page, in other words, the item this page starts with. The service then starts by ordering the data and searching for the first item, then uses the second parameter:
```

- The desired page size, the number of entries following it, along with another entity that uses as the cursor of the next page, which is not displayed to the user.
- A `CancallationToken` which can be used to abort the operation.

The callback returns a `PageResult<TCursor, TItem>`, and it contains:

> - An `ImmutableList<T>` containing the items to be displayed in the resulted page.

Reviewer 3: ```suggestion
- An `ImmutableList<T>` containing the items to be displayed on the resulting page.
```

- A `TCursor` that can be of a different type that points to the beginning of the upcoming page.

To demonstrate this we'll stick to the *PaginatedPeopleApp* we've started before.

### Service

To keep the service agnostic, we'll add the following method to the Service:

*IPeopleService*:

```csharp

> ValueTask<(IImmutableList<Person> CurrentPage, Person NextCursor)> GetPeopleAsync(Person? cursor, uint pageSize, CancellationToken ct);

Reviewer 2: Using the type of the items (a `Person` here) as cursor is not really a common pattern. Usually the API returns a dedicated key (e.g. a `string`) with the first page result, then we only have to pass it to fetch the next page.

For the demo here, you could use the id of the last person to not have the same type.

```

> This method has 3 parameters:
> 
> - The first is a `Person` which uses as a cursor to the first page, i.e. the first `Person` of the requested page.
> - The second is a `uint` representing the number of entities to follow the first one. The reason of using a `uint` is that negative values are not expected here.
> - The third one is a `CancellationToken` which can be used to signal the service to abort loading data.
> 
> The method returns a tuple with two components (same as `PageResult`, only keeping the service agnostic of MVUX):
> 
> - An `IImmutableList<Person>` which includes the entities of this page
> - A `Person` which uses as a cursor as the beginning of the next page.
> 
> > [!TIP]  
> > The cursor does not necessarily have to be a `Person` or the data type the collection contains, it can also be an ID or another key of an entity for the service to lookup and return it along with its upcoming entries.
> 
> In the example above we want to override the `ToString` method of `Person` so that we can later order the entities by their string representation:

Reviewer 3: ```suggestion
This method has 3 parameters:

- The first is a `Person` which uses as a cursor to the first page, i.e. the first `Person` of the requested page.
- The second is a `uint` representing the number of entities to follow the first one. The reason for using a `uint` is that negative values are not expected here.
- The third one is a `CancellationToken` which can be used to signal the service to abort loading data.

The method returns a tuple with two components (same as `PageResult`, only keeping the service agnostic of MVUX):

- An `IImmutableList<Person>` which includes the entities of this page
- A `Person` which uses as a cursor at the beginning of the next page.

> [!TIP]  
> The cursor does not necessarily have to be a `Person` or the data type the collection contains, it can also be an ID or another key of an entity for the service to look up and return it along with its upcoming entries.

In the example above we want to override the `ToString` method of `Person` so that we can later order the entities by their string representation:
```

    return (CurrentPage: collection[..^1].ToImmutableList(), NextCursor: collection[^1]);
}
```

### Model

```csharp
public IListFeed<Person> PeopleCursor =>
    ListFeed<Person>.AsyncPaginatedByCursor(
            // starting off with a blank Person, since the person list is to be ordered by name, any valid name will follow.
            firstPage: new Person(string.Empty, string.Empty), 
            // this will be automatically invoked by the ISupportIncrementalLoading the ListView supports
            getPage: async (cursor, desiredPageSize, ct) =>
                {
                    // gets the tuple result from the service
                    var result = await PeopleService.GetPeopleAsync(cursor, PageSize, ct);
                    // converts tuple to PageResult
                    return new PageResult<Person, Person>(result.CurrentPage, result.NextCursor);
                });
```

### View

There's not much needed in the View as this is magically working via the `ISupportIncrementalLoading` implemented with the `ListView`!

```xml
<ListView ItemsSource="{Binding PeopleCursor}" ItemTemplate="{StaticResource PersonDataTemplate}" />
```

> When the program runs, it works like it did in auto incremental loading, except the people list is sorted by full-name:

Reviewer 3: ```suggestion
When the program runs, it works as it did in auto incremental loading, except the people list is sorted by full-name:
```
